desc/compute/_qs.py

```diff
@@ -0,0 +1,243 @@
+"""Compute functions for quasisymmetry objectives, ie Boozer, Two-Term, and Triple Product."""
+
+from desc.backend import jnp, put, sign
+
+from ._core import check_derivs
+from ._field import (
+    compute_magnetic_field_magnitude,
+    compute_covariant_magnetic_field,
+    compute_B_dot_gradB,
+)
+
+
+def compute_boozer_coords(
+    R_lmn,
+    Z_lmn,
+    L_lmn,
+    i_l,
+    Psi,
+    R_transform,
+    Z_transform,
+    L_transform,
+    B_transform,
+    w_transform,
+    iota,
+    data=None,
+):
+    """Compute Boozer coordinates.
+
+    Assumes transform grids are uniform spacing on single flux surface without symmetry.
+
+    Parameters
+    ----------
+    R_lmn : ndarray
+        Spectral coefficients of R(rho,theta,zeta) -- flux surface R coordinate.
+    Z_lmn : ndarray
+        Spectral coefficients of Z(rho,theta,zeta) -- flux surface Z coordiante.
+    L_lmn : ndarray
+        Spectral coefficients of lambda(rho,theta,zeta) -- poloidal stream function.
+    i_l : ndarray
+        Spectral coefficients of iota(rho) -- rotational transform profile.
+    Psi : float
+        Total toroidal magnetic flux within the last closed flux surface, in Webers.
+    R_transform : Transform
+        Transforms R_lmn coefficients to real space.
+    Z_transform : Transform
+        Transforms Z_lmn coefficients to real space.
+    L_transform : Transform
+        Transforms L_lmn coefficients to real space.
+    B_transform : Transform
+        Transforms spectral coefficients of B(rho,theta,zeta) to real space.
+        B_transform.basis should be of type DoubleFourierSeries.
+    w_transform : Transform
+        Transforms spectral coefficients of w(rho,theta,zeta) to real space.
+        w_transform.basis should be of type DoubleFourierSeries.
+    iota : Profile
+        Transforms i_l coefficients to real space.
+
+    Returns
+    -------
+    data : dict
+        Dictionary of ndarray, shape(num_nodes,) of Boozer harmonics.
+
+    """
+    data = compute_magnetic_field_magnitude(
+        R_lmn,
+        Z_lmn,
+        L_lmn,
+        i_l,
+        Psi,
+        R_transform,
+        Z_transform,
+        L_transform,
+        iota,
+        data=data,
+    )
+    # TODO: can remove this call if compute_|B| changed to use B_covariant
+    data = compute_covariant_magnetic_field(
+        R_lmn,
+        Z_lmn,
+        L_lmn,
+        i_l,
+        Psi,
+        R_transform,
+        Z_transform,
+        L_transform,
+        iota,
+        data=data,
+    )
+
+    NFP = w_transform.basis.NFP
+    if not check_derivs("nu", R_transform, Z_transform, L_transform):
+        return data
```

Comment:
If it does not have the required derivatives, should it not either print something or raise an exception? Instead of a user possibly thinking the compute function worked but finding they do not have the data they expect?

Reply:
the ``compute_*`` functions should never be called directly by a user, they're always either wrapped by equilibrium or objective functions which handle which derivatives are needed. In the future this is something we may want to consider though.